api/src/opentrons/protocol_engine/state/_labware_origin_math.py

```diff
@@ -171,6 +216,194 @@ def _get_parent_deck_item_origin_to_child_labware_placement_origin(
+
+
+def _get_spring_force(
+    child_labware: LabwareDefinition3,
+    parent_deck_item: _Labware3SupportedParentDefinition,
+) -> SpringDirectionalForce | None:
+    """Returns whether the parent-child stackup has a spring that affects positioning."""
+    assert parent_deck_item.features.get("slotFootprintAsParent") is not None
+    assert child_labware.features.get("slotFootprintAsChild") is not None
+
+    parent_spring_force = parent_deck_item.features["slotFootprintAsParent"].get(
+        "springDirectionalForce"
+    )
+    child_spring_force = child_labware.features["slotFootprintAsChild"].get(
+        "springDirectionalForce"
+    )
+
+    if parent_spring_force is not None and child_spring_force is not None:
+        if parent_spring_force != child_spring_force:
+            raise ValueError(
+                f"Parent spring force: {parent_spring_force} does not match child spring force: {child_spring_force}"
+            )
+
+    return parent_spring_force or child_spring_force
+
+
+def _parent_origin_to_slot_bottom_center(
+    parent_deck_item: _Labware3SupportedParentDefinition,
+) -> Point:
+    """Returns offset from the parent's origin to the slot's bottom center."""
+    slot_footprint_as_parent = parent_deck_item.features.get("slotFootprintAsParent")
+    assert slot_footprint_as_parent is not None
+
+    x = (
+        slot_footprint_as_parent["frontRight"]["x"]
+        + slot_footprint_as_parent["backLeft"]["x"]
+    ) / 2
+    y = (
+        slot_footprint_as_parent["frontRight"]["y"]
+        + slot_footprint_as_parent["backLeft"]["y"]
+    ) / 2
+    z = slot_footprint_as_parent["z"]
+
+    return Point(x, y, z)
+
+
+def _parent_origin_to_slot_back_left_bottom(
+    parent_deck_item: _Labware3SupportedParentDefinition,
+) -> Point:
+    """Returns offset from the parent origin to slot back left bottom."""
+    slot_footprint_as_parent = parent_deck_item.features.get("slotFootprintAsParent")
+    assert slot_footprint_as_parent is not None
+
+    x = slot_footprint_as_parent["backLeft"]["x"]
+    y = slot_footprint_as_parent["backLeft"]["y"]
+    z = slot_footprint_as_parent["z"]
+
+    return Point(x, y, z)
+
+
+def slot_bottom_center_to_child_origin(
+    child_labware: LabwareDefinition3,
+) -> Point:
+    """Returns offset from the slot's bottom center to the child origin."""
+    slot_footprint_as_child = child_labware.features.get("slotFootprintAsChild")
+    assert slot_footprint_as_child is not None
+
+    x = (
+        slot_footprint_as_child["frontRight"]["x"]
+        + slot_footprint_as_child["backLeft"]["x"]
+    ) / 2
+    y = (
+        slot_footprint_as_child["frontRight"]["y"]
+        + slot_footprint_as_child["backLeft"]["y"]
+    ) / 2
+    z = slot_footprint_as_child["z"]
+
+    return Point(x, y, z) * -1
+
+
+def _slot_back_left_bottom_to_child_origin(
+    child_labware: LabwareDefinition3,
+) -> Point:
+    """Returns offset from the slot's back left bottom the child's origin."""
```

Comment:
Minor suggestion for clarity, feel free to discard it if you prefer the original.

```suggestion
    """Returns offset from a parent slot's back left bottom to the child's origin."""
```